src/app/core/services/localeService.ts

```diff
@@ -1,24 +1,41 @@
 import { Injectable } from '@angular/core';
-import { appConfig } from 'src/app/shared/helpers/config';
+import { appConfig, LanguageConfig } from 'src/app/shared/helpers/config';
+
 
 @Injectable({
   providedIn: 'root'
 })
 export class LocaleService {
 
-  readonly languages?: { tag: string, path: string }[];
-  readonly currentTag?: string;
+  readonly languages: LanguageConfig[];
+  readonly currentTag: string;
+  readonly currentTagError: boolean
 
   constructor() {
     this.languages = appConfig.languages;
-    this.currentTag = this.languages?.find(l => window.location.pathname.endsWith(l.path))?.tag;
+    const fallbackTag = appConfig.production ? '' : 'en';
+    this.currentTag = this.languages.find(l => window.location.pathname.endsWith(l.path))?.tag ?? fallbackTag;
+    this.currentTagError = !this.currentTag;
   }
 
   navigateTo(langTag: string): void {
-    const nextLang = this.languages?.find(l => l.tag === langTag);
-    const currentLang = this.languages?.find(l => l.tag === this.currentTag);
+    const nextLang = this.languages.find(l => l.tag === langTag);
+    const currentLang = this.languages.find(l => l.tag === this.currentTag);
+
     if (!nextLang || !currentLang) throw new Error('Cannot find new or current lang in configured languages');
-    window.location.href = `${window.location.pathname.replace(currentLang.path, nextLang.path)}${window.location.hash}`;
+
+    const pathname = appConfig.production
+      ? this.getNextPathname(currentLang, nextLang)
+      : this.getNextPathnameForDevelopment(nextLang);
+    window.location.href = `${pathname}${window.location.hash}`;
+  }
+
+  private getNextPathnameForDevelopment(nextLang: LanguageConfig): string {
+    return nextLang.path;
+  }
+
+  private getNextPathname(currentLang: LanguageConfig, nextLang: LanguageConfig): string {
+    return window.location.pathname.replace(currentLang.path, nextLang.path);
   }
```

Comment:
This change might not be in the scope of the PR, I'll detail it shortly.

When changing the language, the app is reloaded by replacing in the URL the current tag by the next tag, but in development, the current tag is not in the URL. As a consequence, the URL computed for next tag is exactly the same the current URL, causing the app *not* to be reloaded.

I will need this change for the last step (which is displaying a language mismatch dialog). Since this PR is quite small I assumed it was ok to do it here.

Reply:
This discussion leads to the same as https://github.com/France-ioi/AlgoreaFrontend/pull/440#discussion_r639838165

Reply:
Fixed